enhancements/sig-architecture/70-spread-policy/README.md

```diff
@@ -0,0 +1,288 @@
+	TopologyKeyTypeClaim TopologyKeyType = "Claim"
+	TopologyKeyTypeLabel TopologyKeyType = "Label"
+)
+```
+
+### API Example
+
+```yaml
+apiVersion: cluster.open-cluster-management.io/v1alpha1
+kind: Placement
+metadata:
+  name: demo-placement
+spec:
+  # numberOfClusters is required if spreadConstraints contains
+  # a constraint of Even type.
+  numberOfClusters: 4
+  # Prediates work before the spread constraints to filter
+  # out some undesired topologies.
+  predicates:
+  - # ...
+  spreadPolicy:
+    # spreadConstraints contain two terms, which work
+    # independently and jointly.
+    spreadConstraints:
+    # When multiple terms are used, the user should specify their weights.
+    # The scheduler will consider to first do the spread according to the
+    # terms with higher weight.
+    # In this example, the scheduler will first spread the workload
+    # by providers.
+    # Within each provider, the scheduler then spread the workload by
+    # regions. 
+    - topologyKey: provider
+      topologyKeyType: Label
+      maxSkew: 2
+      weight: 20
+      # maxSkew is omitted as this term is a soft constraint
+    - topologyKey: region
+      topologyKeyType: Label
+      weight: 10
+  # The spread policy has a scoreCoordinate in prioritizerPolicy.
+  # In Additive mode, the weight of SpreadPolicy is 0 by default.
+  # In Exact mode, the weight of SpreadPolicy should be set by users.
+  prioritizerPolicy:
+    mode: Exact
+    configurations:
+      - scoreCoordinate:
+          builtIn: ResourceAllocatableMemory
+      - scoreCoordinate:
+          builtIn: SpreadPolicy
+        weight: 2
+```
+
+### Implementation
+The current scheduling framework contains two types of plugins, i.e., `Filter`s and `Prioritizer`s. 
+`Filter`s are invoked *before* `Prioritizer`s to handle hard constrains, filtering the clusters. 
+Then, the `Prioritizer`s score the clusters.
+To implement `maxSkew`, we need to first score the clusters (i.e., do the spread), then ensure the `maxSkew` is not exceeded by some filter-like operations for hard constraints.
+This cannot be implemented in the current framework.
+Hence, a new plugin type, `Selector`, is proposed to implement hard constrains *after* `Prioritizer`s.
+The `Selector` interface is showed below.
+
+```go
+// Selector defines a selector plugin that selects clusters after scoring them.
+type Selector interface {
+	Plugin
+
+	// Select returns the selected clusters.
+	Select(ctx context.Context, placement *clusterapiv1beta1.Placement, clusters []*clusterapiv1.ManagedCluster, scores map[string]int64) PluginSelectResult
+}
+
+// PluginSelectResult contains the details of a select plugin result.
+type PluginSelectResult struct {
+	// Selected contains the selected ManagedCluster.
+	Selected []*clusterapiv1.ManagedCluster
+	// Err contains the select plugin error message.
+	Err error
+}
+```
+
+Then, a `EvenSpreadSelector` will be implemented as follows to select n clusters.
```

Comment:
Is there a PR showing this algorithm.  It seems plausible in broad strokes, but the details are a little hard in psuedo code.

Reply:
The code is expected to be available in 1~2 weeks.

Reply:
> The code is expected to be available in 1~2 weeks.

checking back to see if there's a WIP/draft branch.

Reply:
> > The code is expected to be available in 1~2 weeks.
> 
> checking back to see if there's a WIP/draft branch.

Sorry for late reply, the draft implementation is available at https://github.com/huaouo/ocm-placement, and there is a demo video at https://www.youtube.com/watch?v=j7ycK5zaaXA.